src/inferencer/constraintStore.ts

```diff
@@ -1,11 +1,90 @@
+import { Type, Variable, Primitive } from '../types'
+
 export const constraintStore = new Map()
 
-export function updateTypeConstraints(newConstraintLhs: number, newConstraintRhs: number | string) {
-  // step 3a. Add new constraint to constraintStore map
-  // e.g. T1 = T2, T2 = number
-  constraintStore.set(newConstraintLhs, newConstraintRhs)
+export function updateTypeConstraints(newConstraintLhs: Type, newConstraintRhs: Type) {
+  // console.log('updateTypeConstraints')
+  // console.log(newConstraintLhs + ' ' + newConstraintRhs)
+  solveConstraint(newConstraintLhs, newConstraintRhs)
+  // if (detectError(constraintStore)) console.log('Error! Rule 9')
+}
+
+function solveConstraint(constraintLhs: Type, constraintRhs: Type) {
+  // check if both key and value are not base types (Rule 1)
+  if (!is_base_type(constraintLhs) && !is_base_type(constraintRhs)) {
```

Comment:
Refactoring the code to guard clauses would be neater than having many if-else statements. 
For example, instead of having an if-else clause for rule 1, put it in a function called `applyRuleOne(constraintLHS, constraintRHS)`.

Then we would have a happy path that looks something like this:
```
applyRuleOne();
applyRuleTwo();
applyRuleThree();
applyRuleFour();
applyRuleFive();
applyRuleSix();
....
applyRuleNine();
```

Reply:
I think this can KIV first if no time